internal/deregistration/deregistration.go

```diff
@@ -0,0 +1,50 @@
+package deregistration
+
+import (
+	"git.sr.ht/~spc/go-log"
+	"github.com/jakub-dzon/k4e-device-worker/internal/configuration"
+	"github.com/jakub-dzon/k4e-operator/models"
+	"github.com/jakub-dzon/k4e-device-worker/internal/workload"
+)
+
+type Deregistration struct {
+	workloads    *workload.WorkloadManager
+	config       *configuration.Manager
+}
+
+func NewDeregistration(workloadsManager *workload.WorkloadManager, configManager *configuration.Manager) (*Deregistration, error) {
+	deregstration := Deregistration{
+		workloads:                   workloadsManager,
+		config:                      configManager,
+	}
+	return &deregstration, nil
+}
+
+func (d *Deregistration) Update(configuration models.DeviceConfigurationMessage) error {
+	err := d.workloads.RemoveAllWorkloads()
```

Comment:
IMO each of the components should be responsible for executing its own de-registration logic; this package should not know what needs to be done in `d.workloads` and `d.config`. 
It should just call `d.workloads.Deregister`, `d.config.Deregister` and these methods should hide package-private complexity.

We can go even further and extend our config Observer interface with `Remove` method that would be called when needed by configuration module.

That would give us following flow:
- `deviceServer.Disconnect` calls `registration.Registration.Deregister`;
- `registration.Registration.Deregister` calls `configuration.Manager.DeleteDeviceConfig`;
- `configuration.Manager.DeleteDeviceConfig` calls `configuration.Observer.Remove` for each registered observer
- each observer performs its own de-registration logic.

@bardielle, @masayag, @pkliczewski  - WDYT?

Reply:
Actually I thought it's better each components will implements its own deregistration step - I already pushed it as a new commit

About the new suggestion flow- I'm not sure Registration is the correct place because the register component dosen't create the things that Deregistration removes

Reply:
I like the idea. If we'd like to have retries for deregistration logic, then it might have to be duplicated in each component.
Also, when deregistering, we need somehow to coordinate the order of the deregistration which might be reversed to the way we register the observers.

Reply:
I agree that it would be cumbersome to manage order of deregistration in observer pattern. 
@bardielle - nicely done! 
I still think of the `registration` package as something that manages all functionality related to device registration (state), which also includes de-registration and `deregistration.Deregistration.Update` should be renamed and moved to `registration.Registration`

Reply:
One thing that I'm not sure about - the order of deregistration for each observer it improtant (cant remove the nftable or the volumes before deleting the workloads.)
How can I manage the dependencies?

Reply:
I think overall structure is good as it is and I don't see value anymore in using observer pattern in deregistration (due to coordination problems that you and Moti pointed out). The only thing I would like to be different is where deregistration logic is implemented (`registration` package instead of `registration`).